probes/cmd/main.go

```diff
@@ -0,0 +1,65 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"flag"
+	"net/http"
+
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"k8s.io/klog"
+	"k8s.io/perf-tests/probes/pkg/ping/client"
+	"k8s.io/perf-tests/probes/pkg/ping/server"
+)
+
+var (
+	metricAddress = flag.String("metric-bind-address", "0.0.0.0:8080", "The address to serve the Prometheus metrics on.")
+	mode          = flag.String("mode", "", "Mode that should be run. Supported values: ping-server, ping-client")
+)
+
+func main() {
+	klog.InitFlags(flag.CommandLine)
+	flag.Parse()
+	verifyFlags()
+
+	klog.Infof("I'm probes.")
+	klog.Infof("Mode is: %s\n", *mode)
+
+	go func() {
+		http.Handle("/metrics", promhttp.Handler())
+		klog.Infof("Serving metrics on %s\n", *metricAddress)
+		klog.Fatal(http.ListenAndServe(*metricAddress, nil))
+	}()
+
+	switch *mode {
+	case "ping-client":
+		pingclient.Run(pingclient.NewDefaultPingClientConfig())
+	case "ping-server":
+		pingserver.Run(pingserver.NewDefaultPingServerConfig())
+	default:
+		klog.Fatalf("Unrecognized mode: '%s'", *mode)
+	}
+}
+
+func verifyFlags() {
+	if *metricAddress == "" {
+		klog.Fatal("--metric-bind-address not set!")
```

Comment:
Do we need metrics for server? It doesn't sound like "fatal" to me..

Reply:
Yes, we use it to determine whether the probe is up and running. See my reply in the other PR, the one which introduces probes framework.

Reply:
Hmm - I'm not huge fan if that. Why can't we set liveness and/or readiness probes on those pods (by exposing dedicated /healthz handler) rather than hacking with /metrics?

Reply:
I didn't know about liveness and readiness probes.
They're nice features, but after reading more about them, I think they solve a bit different problem then I need to solve.

Liveness probe will restart the container if it's not alive, readiness probe will make sure that no traffic goes to the pod via service as long as it's not ready. 
Both of them are something we may want to consider in the future, but for now I need something else, i.e. an easy way of determining if all probes are up, before starting the actual measurement.

I don't see an easy way of achieving that via liveness/readiness probes, other than querying all the probe pods via api and extracting the readiness probe status. 
And even if I do that I still need to wait for the some of the prometheus targets to become ready (e.g. ping-client). 
So, in my opinion, doing everything via prometheus it's just the simplest solution available right now. 

I agree that using prometheus targets to determine whether a pod is ready is a bit hacky, but I think the same applies to using liveness/readiness probes as they were not designed for that purpose. 

LMK WDYT, I'm happy to discuss that in person when I'm in the office tomorrow.

Reply:
in Endpoints object:
https://github.com/kubernetes/api/blob/master/core/v1/types.go#L3795
the Addresses field contain only ready pods - so if you add readiness probe, Endpoints object will not contain not-ready pods.
So you may simply track the Endpoints object contents.

WDYT?

Reply:
As discussed personally, added comments and todos to implement readiness probes and to consider using something more intuitive to check whether probes are ready. 
Thanks!